src/Sarif.ValidationTests/DefaultObjectPopulatingVisitor.cs

```diff
@@ -0,0 +1,293 @@
+    /// </summary>
+    public class DefaultObjectPopulatingVisitor : SarifRewritingVisitor
+    {
+        public delegate object PrimitiveValueBuilder(bool isRequired);
+
+        // The default builders create a placeholder value for all required property primitives
+        // and emits default (null, 0, false, etc.) values for the remainder. As a result, a
+        // default hydration of this class should contain no non-default values for 
+        // JSON primitive-valued properties that aren't required
+        public static IDictionary<Type, PrimitiveValueBuilder> PopulateRequiredPrimitives()
+        {
+            return new Dictionary<Type, PrimitiveValueBuilder>()
+            {
+                { typeof(bool),    (isRequired) => { return isRequired; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 0; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 0; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "Required string value." : null; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.contoso.com") : null; } }
+            };
+        }
+
+        // These builders exhaustively populate the generated SARIF file, including non-required JSON primitives
+        public static IDictionary<Type, PrimitiveValueBuilder> PopulateAllPrimitives()
+        {
+            // The values in this output are designed to allow conformance to the schema. In our schema,.
+            // for example, some integer values must be > 0. And so, out optional integer value is 
+            // written as 1.
+            return new Dictionary<Type, PrimitiveValueBuilder>()
+            {
+                { typeof(bool),    (isRequired) => { return true; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 1; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 1; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "[Required string value]" : "[Optional string value]"; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(1776, 7, 4).ToUniversalTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.value.contoso.com") : new Uri("https://optional.uri.value.contoso.com"); } }
+            };
+        }
+
+        public DefaultObjectPopulatingVisitor(JsonSchema schema, IDictionary<Type, PrimitiveValueBuilder> primitiveValueBuilders = null)
+        {
+            _schema = schema;
+            _typeToPropertyValueConstructorMap = primitiveValueBuilders ?? PopulateRequiredPrimitives();
```

Comment:
>primitiveValueBuilders [](start = 49, length = 22)

This is another non-obvious feature of the current design. It won't be obvious to the reader that omitting the second argument results in populating only the required primitives. Using the "creation method" pattern makes this explicit.

Reply:
will fix it next change

---
In reply to: [229109180](https://github.com/Microsoft/sarif-sdk/pull/1121#discussion_r229109180) [](ancestors = 229109180)